torch/distributed/distributed_c10d.py

```diff
@@ -1028,61 +1018,60 @@ def _new_process_group_helper(
-        if backend_pg is not None:
-            # Process group wrapper initialization for supported PGs when TORCH_DISTRIBUTED_DEBUG is set
-            if backend_str in [Backend.GLOO, Backend.NCCL, Backend.UCC]:
-                # In debug mode and if GLOO is available, wrap in a wrapper PG that
-                # enables enhanced collective checking for debuggability.
-                if get_debug_level() == DebugLevel.DETAIL:
-                    if not _GLOO_AVAILABLE:
-                        logger.info(
-                            """TORCH_DISTRIBUTED_DEBUG was set to DETAIL, but
-                                    GLOO is not available. Build with Gloo to
-                                    create a wrapper process group in debug mode
-                                    to aid collective desynchronization debugging."""
-                        )
-                    else:
-                        backend_pg = _create_process_group_wrapper(
-                            wrapped_pg=backend_pg,
-                            store_prefix=group_name,
-                            store=prefix_store,
-                            rank=group_rank,
-                            world_size=group_size,
-                            timeout=timeout,
-                        )
-
-        print("set backend")
-        pg._set_backend(torch.device(device), backend_pg_type, backend_pg)
-        print(f"finished creating {backend_pg} for device {device}")
+        # Process group wrapper initialization for supported PGs when TORCH_DISTRIBUTED_DEBUG is set
+        if backend_str in [Backend.GLOO, Backend.NCCL, Backend.UCC]:
+            # In debug mode and if GLOO is available, wrap in a wrapper PG that
+            # enables enhanced collective checking for debuggability.
+            if get_debug_level() == DebugLevel.DETAIL:
+                if not _GLOO_AVAILABLE:
+                    logger.info(
+                        """TORCH_DISTRIBUTED_DEBUG was set to DETAIL, but
+                                GLOO is not available. Build with Gloo to
+                                create a wrapper process group in debug mode
+                                to aid collective desynchronization debugging."""
+                    )
+                else:
+                    backend = _create_process_group_wrapper(
+                        wrapped_pg=backend,
+                        store_prefix=group_name,
+                        store=backend_prefix_store,
+                        rank=group_rank,
+                        world_size=group_size,
+                        timeout=timeout,
+                    )
+
+        # only create single backend pg when backend is set to gloo, nccl, mpi, etc.
+        if backend_enum.lower() in Backend.backend_list:
+            for device in backend_config.get_device_backend_map().keys():
+                pg._register_backend(torch.device(device), backend_type, backend)
```

Comment:
nit: if not yet done so, let's create an issue for tracking the move of `_register_backend` to c++ level.